### PR TITLE
C++ bindings (with exceptions)

### PR DESCRIPTION
**Pull request by skimo-openhub**

### CODE REVIEW

**Comment by skimo-openhub** [comment]
On Wed, Nov 22, 2017 at 02:32:32PM +0000, Tobias Grosser wrote:
> tobig commented on this pull request.
> 
> Moving the brace makes the definitions in the header inconsistent with the ones that are auto-generated. It also moves away from the LLVM style bracing, which has been used for the generated bindings. 

I don't have any context here.  Which brace are you talking about?

skimo

File: interface/cpp.cc [hunk 1]
```diff
@@ -706,6 +706,7 @@ void cpp_generator::print_method_impl(ostream &os, const isl_class &clazz,
 	bool has_callback = false;
 
 	print_method_header(os, clazz, method, fullname, false, kind);
+	osprintf(os, "{\n");
```
Comment (tobig) [inline interface/cpp.cc]: I talk of these braces, changed in 

https://github.com/nicolasvasilache/isl/pull/60/commits/4d8055ead491ab795208ede125349dd07ab449d5

No idea how to reference a commit message yet.

**Comment by tobig** [comment]
An alternative design would be to generate a single header which turns exceptions into abort calls when  compiled with "-fno-exceptions".

**Comment by skimo-openhub** [comment]
On Wed, Nov 22, 2017 at 02:32:32PM +0000, Tobias Grosser wrote:
> tobig commented on this pull request.
> 
> Moving the brace makes the definitions in the header inconsistent with the ones that are auto-generated. It also moves away from the LLVM style bracing, which has been used for the generated bindings. 

Are you saying it's inconsistent with the ones generated in print_ptr_impl(?
I can change those too to improve consistency if you want.

I've been flexible in not enforcing isl formatting
in the C++ bindings, but I don't consider LLVM code style the goal here.

In any case, as the commit message says, this is a side-effect
of a clean-up.

skimo

**Comment by skimo-openhub** [comment]
On Wed, Nov 22, 2017 at 02:52:19PM +0000, Tobias Grosser wrote:
> An alternative design would be to generate a single header which turns exceptions into abort calls when  compiled with "-fno-exceptions".

That would mean you wouldn't be able to recover from errors
(including compute-outs).  That doesn't sound very useful.
In any case, this is a design decision for the bindings
_without_ exceptions.

skimo

**Comment by skimo-openhub** [comment]
On Wed, Nov 22, 2017 at 02:41:30PM +0000, Tobias Grosser wrote:
> tobig commented on this pull request.
> 
> Is there any compatibility story between these bindings and the new (with exception) bindings?

Do you mean the ones _without_ exceptions?

> It seems to me that any code developed on top of these bindings will not be usable in the context of LLVM? Is this right?

That depends on LLVM.  If you can't use exceptions in LLVM,
then you won't be able to use the isl bindings with exceptions.

skimo

**Comment by tobig** [comment]
> Do you mean the ones __without__ exceptions?

Yes.

> That depends on LLVM.  If you can't use exceptions in LLVM, then you won't be able to use the isl bindings with exceptions.

Right. Hence, my question. I feel that any code developed with these bindings would not work in LLVM. This will make moving code into or out of LLVM close to impossible. Specifically, ppcg++ cannot be shared with LLVM.


I wonder if

**Comment by tobig** [comment]
(No idea how to cite here)

I talked about the braces in isl.h.top. I don't care very much about the location of braces, but at least ATM they are consistent. Feel free to ignore my comment. Just wanted to point this out.

**Comment by skimo-openhub** [comment]
On Wed, Nov 22, 2017 at 05:04:21PM +0000, Tobias Grosser wrote:
> (No idea how to cite here)
> 
> I talked about the braces in isl.h.top. I don't care very much about the location of braces, but at least ATM they are consistent. Feel free to ignore my comment. Just wanted to point this out.

Hmm...
I assume you mean isl.h.pre.  I only see two out-of-line
method definitions (exception::exception and exception::create) and
they have the opening brace on the next line.

skimo

**Pull request merged**